### PR TITLE
Add lora weights to state dict conversions

### PR DESCRIPTION
**Pull request by lucylq**
#### Context
What is the purpose of this PR? Is it to
- [ x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* Add LoRA weights to Meta<->torchtune state dict conversions

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ x] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

**Review by joecummings — APPROVED** [review]

**Comment by ebsmothers** [comment]
Hi @lucylq can you share a bit more context on this? Most of our state dict mappings do not require LoRA keys because we merge the LoRA weights back into the original model after training. And if you are trying to load the adapter weights separately from e.g. a model in the Meta format trained with torchtune, they should already be saved in torchtune's format ([ref](https://github.com/pytorch/torchtune/blob/0ddd4b93c83de60656fb3db738228b06531f7c1e/torchtune/training/checkpointing/_checkpointer.py#L1225-L1235))

**Comment by lucylq** [comment]
> Hi @lucylq can you share a bit more context on this? Most of our state dict mappings do not require LoRA keys because we merge the LoRA weights back into the original model after training. And if you are trying to load the adapter weights separately from e.g. a model in the Meta format trained with torchtune, they should already be saved in torchtune's format ([ref](https://github.com/pytorch/torchtune/blob/0ddd4b93c83de60656fb3db738228b06531f7c1e/torchtune/training/checkpointing/_checkpointer.py#L1225-L1235))

Hi @ebsmothers yes for sure. I'm trying to enable LoRA as a use-case for program-data sharing in executorch. We'd like for separate LoRA models to share the same foundation weights to reduce on-device memory usage; this means we can't merge the lora weights in. The final file structure would look something like this:

- lora1.pte: model + lora weights
- lora2.pte: model + lora weights
- foundation.ptd: foundation weights only.

I'm currently using adapter weights from torchtune, and want to convert them back to meta, so we can use it with the meta llama checkpoint if that's what a user ends up having. I'd like to use torchtune's convert function, but I can also create a copy in the executorch repo if that's preferred?

**Comment by ebsmothers** [comment]
Thanks @lucylq for the additional info. Given that we only call tune_to_meta/meta_to_tune in torchtune on base checkpoints (or checkpoints whose LoRA weights have already been merged back into the original model), I think this change should be fine from torchtune's save/load point of view. So this looks good to me.
